plutus-ledger-api/src/PlutusLedgerApi/Common/Eval.hs

```diff
@@ -250,9 +255,17 @@ evaluateScriptCounting ll pv verbose ectx p args = swap $ runWriter @LogOutput $
     let (res, UPLC.CountingSt final, logs) =
             evaluateTerm UPLC.counting pv verbose ectx appliedTerm
     tell logs
-    liftEither $ first CekError $ void res
+    case res of
+        Left e -> throwError (CekError e)
+        Right v -> unless (ll == PlutusV1 || ll == PlutusV2 || isUnit v) $ throwError NonUnitReturnValue
     pure final
 
+isUnit :: UPLC.Term UPLC.NamedDeBruijn DefaultUni DefaultFun () -> Bool
+isUnit t = case readKnownConstant t of
```

Comment:
I think you can simply use `readKnown`? It should figure out that you want a constant.

Also, call it `isBuiltinUnit`?